### ISSUE TITLE
Add pixel perfect collision detection

### ISSUE THREAD

**Issue by ericclack** [lordmauve/pgzero#216]
Hi, many thanks for creating Pgzero, it's working really well at our Coder Dojo in Brighton :)

I've hacked together some code for pixel perfect collision detection, would you consider adding it to pgzero? I can submit a pull request if you are happy to review. 

Here's the code:

```
def collide_pixel(actor1, actor2):

    # Get masks for pixel perfect collision detection:
    for a in [actor1, actor2]:
        if not hasattr(a, 'mask'):
            a.mask = mask.from_surface(a._orig_surf)

    # Offsets based on current positions of actors
    xoffset = int(actor2.left - actor1.left)
    yoffset = int(actor2.top - actor1.top)

    # Check for overlap = collision
    return actor1.mask.overlap(actor2.mask, (xoffset, yoffset))
```

Bit more background here:
http://chalkpath.co.uk/blog/html/2020/01/12/pixel_perfect_collision_detection_with_pygamezero.html

**Comment by lordmauve** [comment]
I don't think pixel perfect collision detection by checking mask intersections is a good idea to put into Pygame Zero. It's not a good lesson to teach because it's too slow for general purpose use in games. That's not to say it has no uses at all. It vectorises well, and could be GPU accelerated (but not in Pygame). However, the point is that you'd have to carefully choose where to apply it to get the benefit of it.

Perhaps Pygame Zero should offer collision detection features, but I'd want it to be via quadtrees and polygon intersection tests, not pixel masks.

**Comment by ericclack** [comment]
Thanks for your feedback. 

Here's one of the use-cases for better-than-rect collision detection: a maze game where the coder draws the walls and the whole maze is rendered as a single image:
https://scratch.mit.edu/projects/11710850/editor/

In this case the coder doesn't use mask collision detection but something more like 'pixel colour at location'.

What are your thoughts on how to achieve this with Pgzero? I can see that we could get the coder to draw the maze with individual sprites for each wall and use colliderect on them, but that would be a lot of work.

**Comment by lordmauve** [comment]
In a maze with walls like that, you don't really want free movement, because that will create quite a fiddly experience.

Maybe I should write an example game to show how best to write something like that.

**Comment by ericclack** [comment]
Well I'd be happy to explore too. If you want to give me some pointers, things you'd explore first, I'll have a go.

**Comment by lordmauve** [comment]
I wrote an example game; let me know what you think:

https://github.com/lordmauve/pgzero/tree/master/examples/maze/

**Comment by ericclack** [comment]
That's an interesting example, I like the maze generation code and I can see how you've redone the collision detection to avoid moves that are not possible.

Going back to the original question I asked "can we have pixel perfect collision detection in Pgzero?" I think the answer is: you probably don't need it, generate the world using sprites or data structures so that you can check for collisions in a way that's more efficient than checking pixels. 

If we do decide to use it occasionally do you think my method `collide_pixel` above is relatively safe? It is poking about in the Pygame internals, e.g. using `_orig_surf`, which is why I ask.

**Comment by lordmauve** [comment]
> If we do decide to use it occasionally do you think my method collide_pixel above is relatively safe? It is poking about in the Pygame internals, e.g. using _orig_surf, which is why I ask.

No, you're right `_orig_surf` might change. However, you can always get a surface with `images.load(name)`. So if the Actor's `.image` attribute is a string, it would be

```
mask.from_surface(images.load(actor.image))
```

> Going back to the original question I asked "can we have pixel perfect collision detection in Pgzero?" I think the answer is: you probably don't need it, generate the world using sprites or data structures so that you can check for collisions in a way that's more efficient than checking pixels.

Exactly. I think we probably need a documentation section that talks about game maths, algorithms and datastructures, including collision detection. It's on my to-do list.

**Comment by Mambouna** [comment]
@lordmauve Hi, has your thinking changed on this in the meantime? I just looked through the thread and I think both could / should be possible. I agree that it's quite unperformant but I'm not sure that it's necessarily teaching the wrong lesson to use mask collision detection sparingly. In regard to teachability, the concept is easy to explain and it comes up quite quickly when using PGZero in class since students notice very soon that the collisions they are detecting sometimes don't actually "hit".

I do think that ultimately, a more robust solution would be good. Looking towards a perfect solution when it's unlikely to come anytime soon shouldn't block a "good enough" or "good start" solution I think.

**Comment by lordmauve** [comment]
Yeah, maybe. Now I think about it the cost of each mask test is equal to or less than a blit of the intersecting area. The issue with all collision detection is that there can be a lot of pairs (`n * (n - 1) / 2`) while blits are linear.

Also the PyGame docs give examples like

```python
dx = mask.overlap_area(other, (x + 1, y)) - mask.overlap_area(other, (x - 1, y))
dy = mask.overlap_area(other, (x, y + 1)) - mask.overlap_area(other, (x, y - 1))
```

which are good ways to increase the cost.

One approach would be to add a method but document that it could be expensive if checking large numbers of pairs with large sprites.

I think I still prefer a holistic solution to collision detection with a broad phase and computed collision volumes. I think beginners would get more solid results with less frustration.

**Comment by Mambouna** [comment]
If I understand the code example correctly, that's used to get the direction component of a collission, correct? Would you want any solution to include the option to get the directionality of a collision? So far it hasn't been part of the discussion I think. If you only prefer a bigger all-round solution, then that's that, just thought I'd bump the discussion since I think it might be useful. In case you do consider a function, I can set something up soon.

Here's a first idea of how that could be phrased in the docs (starting from the par about .colliderect()):

```rst
Actors have all the same attributes and methods as :ref:`Rect <rect>`,
including methods like `.colliderect()`__ which can be used to test whether
two actors have collided. This is quick but imperfect collision detection.

.. __: https://www.pygame.org/docs/ref/rect.html#pygame.Rect.colliderect

Additionally, collisions between actors can be checked more precisely by
calling ``actor1.collidemask(actor2)``. This checks collision down to the
pixel level, meaning that if the rects of two actors overlap but their
images don't actually intersect, a collision won't be reported. This is
a lot more precise but also more work to check for. If coarse detection 
is fine, always use ``.colliderect()``. If you need high precision, use
``.collidemask()`` only where it is necessary.
```
